### PR TITLE
[PDI-14958] - Grouping in Transformation Executor stopped working

### PR DESCRIPTION
**Pull request by mchen-len-son**
- fixed job/trans execution grouping by field and by size
- added unit tests

@brosander, @akhayrutdinov, please review the changes.

### CODE REVIEW

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Dpentaho.coding.standards.version=1.0.4 -Djunit.jvmargs=-XX:PermSize=256m -f engine/build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  engine/src/org/pentaho/di/trans/steps/jobexecutor/JobExecutor.java
  engine/src/org/pentaho/di/trans/steps/transexecutor/TransExecutor.java
  engine/test-src/org/pentaho/di/trans/steps/jobexecutor/JobExecutorTest.java
  engine/test-src/org/pentaho/di/trans/steps/transexecutor/TransExecutorUnitTest.java










Unit test coverage change:
============
org.pentaho.di.trans.Trans:
 * instructionPercentCoverageChange : +0.14%
 * branchPercentCoverageChange : +0.10%
 * linePercentCoverageChange : +0.34%
 * complexityPercentCoverageChange : +0.42%
 * methodPercentCoverageChange : +1.44%
    
org.pentaho.di.trans.steps.uniquerows.UniqueRowsMeta:
 * instructionPercentCoverageChange : +0.17%
    
org.pentaho.di.trans.steps.jobexecutor.JobExecutorData:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.jobexecutor.JobExecutorMeta.new Object() {...}:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :88.46%
    
org.pentaho.di.trans.steps.jobexecutor.JobExecutorMeta:
 * instructionPercentCoverageChange : +0.88%
 * linePercentCoverageChange : +1.65%
 * complexityPercentCoverageChange : +3.98%
 * methodPercentCoverageChange : +6.73%
    
org.pentaho.di.trans.steps.sortedmerge.SortedMergeMeta:
 * instructionPercentCoverageChange :**-0.18%**
    
org.pentaho.di.core.util.CurrentDirectoryResolver:
 * branchPercentCoverage :8.97%
 * linePercentCoverage :25.00%
 * methodPercentCoverage :60.00%
 * complexityPercentCoverage :6.82%
 * instructionPercentCoverage :14.29%
    
org.pentaho.di.trans.steps.jobexecutor.JobExecutor:
 * branchPercentCoverage :28.18%
 * linePercentCoverage :39.59%
 * methodPercentCoverage :50.00%
 * complexityPercentCoverage :20.90%
 * instructionPercentCoverage :31.80%
    
org.pentaho.di.trans.steps.randomvalue.RandomValueMeta:
 * instructionPercentCoverageChange :**-0.15%**
 * branchPercentCoverageChange :**-2.08%**
 * complexityPercentCoverageChange :**-2.13%**
    
org.pentaho.di.trans.steps.jobexecutor.JobExecutorParameters:
 * instructionPercentCoverageChange : +1.89%
 * linePercentCoverageChange : +3.28%
 * complexityPercentCoverageChange : +11.11%
 * methodPercentCoverageChange : +14.29%
    
org.pentaho.di.trans.steps.transexecutor.TransExecutor:
 * instructionPercentCoverageChange : +6.63%
 * branchPercentCoverageChange : +13.21%
 * linePercentCoverageChange : +5.19%
 * complexityPercentCoverageChange : +13.54%
 * methodPercentCoverageChange :**-3.51%**
    
org.pentaho.di.job.entries.sql.JobEntrySQL:
 * instructionPercentCoverageChange :**-1.64%**
 * linePercentCoverageChange :**-1.97%**

**Comment by mchen-len-son** [comment]
@akhayrutdinov, I've replaced package-level methods comments with @VisibleForTesting.
Recheck, please.

File: engine/src/org/pentaho/di/trans/steps/jobexecutor/JobExecutor.java [hunk 1]
```diff
@@ -128,38 +138,40 @@ public boolean processRow( StepMetaInterface smi, StepDataInterface sdi ) throws
...
+        }
+      }
+
+      // Add next value AFTER job execution, in case we are grouping by field (see PDI-14958),
+      // and BEFORE checking size of a group, in case we are grouping by size (see PDI-14121).
       data.groupBuffer.add( new RowMetaAndData( getInputRowMeta(), row ) ); // should we clone for safety?
+
+      // Grouping by size.
+      // If group buffer size exceeds specified limit, then execute job and flush group buffer.
       if ( data.groupSize >= 0 ) {
```
Comment (akhayrutdinov) [inline engine/src/org/pentaho/di/trans/steps/jobexecutor/JobExecutor.java]:
```
      if ( data.groupSize >= 0 ) {
        // Pass the input rows in blocks to the job result rows...
        if ( data.groupSize != 0 ) { 
...
        }
      }
```
I think, the internal block can be removed by restricting external's condition

File: engine/src/org/pentaho/di/trans/steps/transexecutor/TransExecutor.java [hunk 2]
```diff
@@ -91,32 +98,37 @@ public boolean processRow( StepMetaInterface smi, StepDataInterface sdi ) throws
...
+        }
+      }
+
+      // Add next value AFTER transformation execution, in case we are grouping by field (see PDI-14958),
+      // and BEFORE checking size of a group, in case we are grouping by size (see PDI-14121).
       transExecutorData.groupBuffer.add( new RowMetaAndData( getInputRowMeta(), row ) ); // should we clone for safety?
+
+      // Grouping by size.
+      // If group buffer size exceeds specified limit, then execute transformation and flush group buffer.
       if ( transExecutorData.groupSize >= 0 ) {
```
Comment (akhayrutdinov) [inline engine/src/org/pentaho/di/trans/steps/transexecutor/TransExecutor.java]: The same is here: internal `if` block seems to be redundant

**Comment by mchen-len-son** [comment]
@akhayrutdinov, commited changes due to your comments. Recheck, please.

**Comment by akhayrutdinov** [comment]
Checkstyle is OK.
Tests (both) are passing.
Merged.

**Pull request merged**